### DISCUSSION TITLE
Wait for step to complete execution before jumping into next step

### DISCUSSION THREAD

**Discussion by avinashk14** [badeball/cypress-cucumber-preprocessor#1071] [Q&A]
Hello,

I've been trying for one step definition to complete its execution (along with all its back-end calls) before jumping into the next step definition and executing its line, any help is greatly appreciated.

Example Scenario:
```
Scenario: Test scenario
    Given I create a resource
    Then I should have the resource
```

Step definitions:
```
Given('I create a resource', function() {
  setTimeout(() => {
    console.log("I create a resource");
  }, 1000);
})

Then('I should have the resource', function() {
  console.log('I should have the resource')
})
```

What I am looking for is for the step 'I create a resource' to wait for all its lines executed and then the next step 'I should have the resource' executes.
From the console log, I expect to see: "I create a resource", " I should have the resource"; what I actually see: "I should have the resource", "I create a resource".

Is there a way for the step definition to wait? - Reason for asking is because we have a lot of scenarios where resource creation/fetch is the first step and the subsequent steps check on the resource, but they fail because the create/fetch step doesn't wait for the back-end call to finish.
(I'm at cypress v12.17.4 and cypress-cucumber-preprocessor v18.0.4)

**Comment by badeball** [comment]
Use Cypress commands for resource creation. IE. `cy.request(..)` and you'll see that steps are executed consecutively. Alternatively you can wrap your code in promises, https://docs.cypress.io/api/utilities/promise.

**Reply by avinashk14** [reply]
I am using Cypress commands for resource create/fetch, but still the steps are not waiting for execution to complete. 
Ex, for post I have: 
```
    async post(path, body = {}) {
        const args = {
            requestType: 'POST',
            path: path,
            contentType: 'application/json'
        };
        return new Cypress.Promise((resolve, reject) => {
            cy.request({
                method: args.requestType,
                url: this.url.concat(path),
                headers: this.header_options(args),
                body: body,
                failOnStatusCode: false
            }).then(response => resolve(response))
        })
    }
```

From how I understand, the next step would only execute once the previous step completes its entire execution, or the current step would not pass execution to next step if there are any async calls are still pending. Please correct me if I'm wrong.

**Reply by badeball** [reply]
I can't verify or correct your understanding in this discussion. You need to provide a complete, runable and minimal example that illustrates the behavior you consider surprising.

**Reply by avinashk14** [reply]
Here is a runable example

```
Feature: Test Feature
  Scenario: Test scenario
    When I create a resource
    Then I should have the resource
```

```
const { Then, When } = require('@badeball/cypress-cucumber-preprocessor')

When('I create a resource', function() {
  post().then((res) => {
    cy.wrap(res.body.message).as('message')
  })
})

Then('I should have the resource', function() {
  expect(this.message).to.eq('Created')
})

async function post() {
  return new Cypress.Promise((resolve, reject) => {
      cy.request({
          method: 'POST',
          url: 'https://19lq5.wiremockapi.cloud/test_post'
      }).then(response => resolve(response))
  })
}
```
It fails on the expectation "expected undefined to equal **Created**"

**Reply by badeball** [reply]
Your example passes as far as I can see.

![2023-08-17-195725_5120x1440_scrot](https://github.com/badeball/cypress-cucumber-preprocessor/assets/730419/a1ab7eba-5190-42e5-baba-689c513d22a5)

**Reply by badeball** [reply]
You also don't need a wrapper-promise in this case, the following is fine.

```js
When('I create a resource', function() {
  cy.request({
    method: 'POST',
    url: 'https://19lq5.wiremockapi.cloud/test_post'
  }).then((res) => {
    cy.wrap(res.body.message).as('message')
  })
})

Then('I should have the resource', function() {
  expect(this.message).to.eq('Created')
})
```

**Reply by avinashk14** [reply]
Thanks, I will try to work around your suggestion.

**Reply by avinashk14** [reply]
To note: the runable example mentioned above passes on Chrome v115, but fails on Electron v106 and Firefox v116, not very sure why it behaves differently on different browsers.

**Reply by badeball** [reply]
Remove `async` from your function, it's not really asynchronous in the traditional sense.
